compiler/rustc_middle/src/values.rs

```diff
@@ -154,13 +154,37 @@ impl<'tcx, T> Value<TyCtxt<'tcx>> for Result<T, &'_ ty::layout::LayoutError<'_>>
             } else {
                 tcx.def_span(def_id)
             };
-            struct_span_err!(tcx.sess, span, E0733, "recursion in an `async fn` requires boxing")
-                .span_label(span, "recursive `async fn`")
-                .note("a recursive `async fn` must be rewritten to return a boxed `dyn Future`")
-                .note(
-                    "consider using the `async_recursion` crate: https://crates.io/crates/async_recursion",
-                )
-                .emit()
+            let mut diag = struct_span_err!(
+                tcx.sess,
+                span,
+                E0733,
+                "recursion in an `async fn` requires boxing"
+            );
+            diag.note("a recursive `async fn` call must introduce indirection such as `Box::pin` to avoid an infinitely sized future");
+            diag.note(
+                "consider using the `async_recursion` crate: https://crates.io/crates/async_recursion",
```

Comment:
do we still want to suggest the `async_recursion` crate. manually changing the recursive call to `Box::pin(recur()).await)` feels preferable to adding a new dependency. In my experience suggesting multiple solutions in diagnostics causes them to be less helpful as they end up somewhat overwhelming the user, causing them to ignore the diagnostic entirely 🤔

Reply:
@rust-lang/wg-async relevant for you: if simply using `Box::pin` works now, should we still suggest to use the `async_recursion` macro?